src/wms/src/main/java/org/geoserver/wms/legendgraphic/JSONLegendGraphicBuilder.java

```diff
@@ -0,0 +1,843 @@
+            }
+            ret.element(STROKE_DASHARRAY, dArray);
+            Expression dashOffset = stroke.getDashOffset();
+            if (dashOffset != null) {
+                ret.element(STROKE_DASHOFFSET, dashOffset.evaluate(feature));
+            }
+        }
+        return ret;
+    }
+
+    /**
+     * @param symbolizer
+     * @return
+     */
+    private JSONObject processSymbolizer(Symbolizer symbolizer) {
+        JSONObject ret = new JSONObject();
+        String name = symbolizer.getName();
+        if (name != null && !name.isEmpty()) {
+            ret.element(NAME, name);
+        }
+        Unit<Length> uom = symbolizer.getUnitOfMeasure();
+        if (uom != null) {
+            ret.element(UOM, uom.toString());
+        }
+        Description desc = symbolizer.getDescription();
+        if (desc != null) {
+            InternationalString title = desc.getTitle();
+            if (title != null) {
+                ret.element(TITLE, title.toString());
+            }
+            InternationalString abs = desc.getAbstract();
+            if (abs != null) {
+                ret.element(ABSTRACT, abs.toString());
+            }
+        }
+        Expression geometry = symbolizer.getGeometry();
+        if (geometry != null) {
+            ret.element(GEOMETRY, printExpression(geometry, Geometry.class));
+        }
+        if (symbolizer instanceof PointSymbolizer) {
+            ret = processPointSymbolizer(ret, (PointSymbolizer) symbolizer);
+        } else if (symbolizer instanceof LineSymbolizer) {
+            ret = processLineSymbolizer(ret, (LineSymbolizer) symbolizer);
+        } else if (symbolizer instanceof PolygonSymbolizer) {
+            ret = processPolygonSymbolizer(ret, (PolygonSymbolizer) symbolizer);
+        } else if (symbolizer instanceof RasterSymbolizer) {
+            ret = processRasterSymbolizer(ret, (RasterSymbolizer) symbolizer);
+        } else if (symbolizer instanceof TextSymbolizer) {
+            ret = processTextSymbolizer(ret, (TextSymbolizer) symbolizer);
+        }
+
+        Map<String, String> opts = symbolizer.getOptions();
+        ret = processVendorOptions(ret, opts);
+        return ret;
+    }
+
+    /**
+     * @param ret
+     * @param opts
+     */
+    private JSONObject processVendorOptions(JSONObject ret, Map<String, String> opts) {
+        if (!opts.isEmpty()) {
+            JSONObject vendorOpts = new JSONObject();
+
+            for (Entry<String, String> opt : opts.entrySet()) {
+                vendorOpts.element(opt.getKey(), opt.getValue());
+            }
+            ret.element(VENDOR_OPTIONS, vendorOpts);
+        }
+        return ret;
+    }
+
+    /**
+     * @param ret
+     * @param symbolizer
+     * @return
+     */
+    private JSONObject processTextSymbolizer(JSONObject ret, TextSymbolizer symbolizer) {
+        ret.element(LABEL, printExpression(symbolizer.getLabel(), String.class));
+        JSONArray fonts = new JSONArray();
```

Comment:
I don't see it encoding the font fill, what if fonts have color?

Reply:
Also, the font graphic (shield) encoding appears to be missing, you'll have to check if you can cast to TextSymbolizer2 to handle it.